contracts-proc/src/interface_id.rs

```diff
@@ -0,0 +1,122 @@
+        };
+
+        let mut override_name = None;
+        for attr in mem::take(&mut func.attrs) {
+            let Some(ident) = attr.path().get_ident() else {
+                func.attrs.push(attr);
+                continue;
+            };
+            if *ident == "selector" {
+                if override_name.is_some() {
+                    error!(attr.path(), "more than one selector attribute");
+                }
+                let args: SelectorArgs = match attr.parse_args() {
+                    Ok(args) => args,
+                    Err(error) => error!(ident, "{}", error),
+                };
+                override_name = Some(args.name);
+                continue;
+            }
+            func.attrs.push(attr);
+        }
+
+        let sol_name = override_name
+            .unwrap_or_else(|| func.sig.ident.to_string().to_case(Case::Camel));
+
+        let arg_types: Vec<_> = func
+            .sig
+            .inputs
+            .iter()
+            .filter_map(|arg| match arg {
+                FnArg::Typed(t) => Some(t.ty.clone()),
+                FnArg::Receiver(_) => None,
+            })
+            .collect();
+
+        let selector = quote! { u32::from_be_bytes(stylus_sdk::function_selector!(#sol_name #(, #arg_types )*)) };
+        selectors.push(selector);
+    }
+
+    let name = input.ident.clone();
+    let vis = input.vis.clone();
+    let attrs = input.attrs.clone();
+    let trait_items = input.items.clone();
+    let (_impl_generics, ty_generics, where_clause) =
+        input.generics.split_for_impl();
+
+    quote! {
+        #(#attrs)*
+        #vis trait #name #ty_generics #where_clause {
+            #(#trait_items)*
+
+            /// Solidity interface id associated with current trait.
+            const INTERFACE_ID: u32 = {
+                #(#selectors)^*
+            };
+        }
+    }
+    .into()
+}
+
+struct SelectorArgs {
+    name: String,
+}
+
+impl Parse for SelectorArgs {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let mut name = None;
+
+        if input.is_empty() {
+            error!(@input.span(), "missing id or text argument");
+        }
+
+        while !input.is_empty() {
+            let ident: Ident = input.parse()?;
+            let _: Token![=] = input.parse()?;
+
+            match ident.to_string().as_str() {
+                "name" => {
+                    let lit: LitStr = input.parse()?;
+                    if name.is_some() {
+                        error!(@lit, r#"only one "name" is allowed"#);
+                    }
+                    name = Some(lit.value());
+                }
+                _ => error!(@ident, "Unknown selector attribute"),
+            }
+
+            // allow a comma
+            let _: Result<Token![,]> = input.parse();
+        }
+
+        if let Some(name) = name {
+            Ok(Self { name })
+        } else {
+            error!(@input.span(), r#""name" is required"#);
+        }
+    }
+}
```

Comment:
Maybe this is a dumb question, but if we are only expecting the "name" attribute when parsing can't we just rely on `syn` parsing and do something simpler like this?
```
impl Parse for SelectorArgs {
    fn parse(input: ParseStream) -> Result<Self> {
        let ident: Ident = input.parse()?;
        let _: Token![=] = input.parse()?;
        let lit: LitStr = input.parse()?;

        if ident != "name" {
            return Err(syn::Error::new_spanned(ident, "expected 'name'"));
        }

        Ok(SelectorArgs { name: lit.value() })
    }
}
```

I run the unit tests and all are passing with this.

Reply:
`selector` attribute parsing was copied from stylus-sdk :)
Just errors are better there.. 
e.g if you would write `#[selector(selector ! "mySelector")]` the first error your version would show is `expected '='`. Or if there are some additional tokens just general error.
But I agree it can be refactored!